### PR TITLE
fix: improve backend autocomplete

### PR DESCRIPTION
**Pull request by rephus**
Closes:  https://github.com/lightdash/lightdash/issues/12789

## Changes: 

- cache entire hashed query, this includes user attributes (if any) . Note, this is secure, but less efficient,s ince we need to get all metadata from explore and metricquery to compute the query to get the cache key to know if it is cached or not;
- use s3cacheClient to store results in s3, rather than in memory
- change in env variables  

### Description:

Same UI: 

[Screencast from 2024-12-11 10-04-29.webm](https://github.com/user-attachments/assets/fdbf3eaa-9837-4844-8755-db19237d5651)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 1]
```diff
@@ -2616,6 +2599,42 @@ export class ProjectService extends BaseService {
...
+            .createHash('sha256')
+            .update(queryHashKey)
+            .digest('hex');
+
+        const isCacheEnabled =
+            this.lightdashConfig.resultsCache.autocompleteEnabled &&
+            this.s3CacheClient.isEnabled();
+
+        if (!forceRefresh && isCacheEnabled) {
+            const isCached = await this.s3CacheClient.getResultsMetadata(
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:2615]: getResultsMetadata does not throw an error if results does not exist, so this needs to be called before getResults. It works like `hasResults`

File: packages/frontend/src/components/common/Filters/FilterInputs/FilterStringAutoComplete.tsx [hunk 2]
```diff
@@ -161,7 +161,7 @@ const FilterStringAutoComplete: FC<Props> = ({
                     <Tooltip
                         withinPortal
                         position="left"
-                        label={`Click here to cache results now`}
+                        label={`Click here to refresh cache filter values`}
```
Comment (rephus) [inline packages/frontend/src/components/common/Filters/FilterInputs/FilterStringAutoComplete.tsx:164]: quick small copy change suggested by Jake. More UI changes will be added on a separate PR

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 3]
```diff
@@ -2616,6 +2599,42 @@ export class ProjectService extends BaseService {
             this.lightdashConfig.query.timezone || 'UTC',
         );
 
+        const queryHashKey = metricQuery.timezone
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:2602]: same hashkey logic as results

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 4]
```diff
@@ -2615,6 +2598,42 @@ export class ProjectService extends BaseService {
...
+
+            if (isCached !== undefined) {
+                const cacheEntry = await this.s3CacheClient.getResults(
+                    queryHash,
+                );
+                const stringResults =
+                    await cacheEntry.Body?.transformToString();
+                if (stringResults) {
+                    try {
+                        return JSON.parse(stringResults);
```
Comment (ZeRego) [inline packages/backend/src/services/ProjectService/ProjectService.ts:2645]: Response doesn't have`cached: true,` anymore. is that expected?

Comment (rephus) [reply packages/backend/src/services/ProjectService/ProjectService.ts:2645]: When we upload the results to s3, we include all the metadata , like cached, or refreshedAt. When  fetching from s3, we simply return the same object we previously store. 

To avoid conflicts with normal cached results,  thes are stored in a different cacheKey

**Review by ZeRego — APPROVED** [review]
